### PR TITLE
impl GetTypeRegistration and FromType for Vec<T>

### PR DESCRIPTION
**Pull request by CleanCut**
See https://github.com/bevyengine/bevy/issues/1528#issuecomment-788130786

### CODE REVIEW

File: crates/bevy_reflect/src/impls/std.rs [hunk 1]
```diff
@@ -112,6 +112,20 @@ impl<T: Reflect> Reflect for Vec<T> {
     }
 }
 
+impl<T: Reflect + Clone> GetTypeRegistration for Vec<T> {
+    fn get_type_registration() -> TypeRegistration {
+        let mut registration = TypeRegistration::of::<Vec<T>>();
+        registration.insert::<Vec<T>>(FromType::<Vec<T>>::from_type());
+        registration
+    }
+}
```
Comment (Davier) [inline crates/bevy_reflect/src/impls/std.rs:121]: The "insert" line comes from `#reflect(Deserialize)`. It's not needed for `Serialize`, `Hash` and `PartialEq`, only for others (`Deserialize`, `Component`, `MapEntity` and any other reflected trait).
```suggestion
impl<T: Reflect + Clone> GetTypeRegistration for Vec<T> {
    fn get_type_registration() -> TypeRegistration {
        let mut registration = TypeRegistration::of::<Vec<T>>();
        registration.insert::<ReflectDeserialize>(FromType::<Vec<T>>::from_type());
        registration
    }
}
```

File: crates/bevy_reflect/src/impls/std.rs [hunk 2]
```diff
@@ -112,6 +112,20 @@ impl<T: Reflect> Reflect for Vec<T> {
...
+
+impl<T> FromType<Vec<T>> for Vec<T> {
+    fn from_type() -> Self {
+        vec![]
+    }
+}
+
```
Comment (Davier) [inline crates/bevy_reflect/src/impls/std.rs:128]: `FromType` is already implemented for the types that need it (like `ReflectDeserialize`).
```suggestion
```

**Comment by Davier** [comment]
Why did you add the `Clone` bound?

File: crates/bevy_reflect/src/impls/std.rs (hunk 1 above)
Comment (CleanCut) [reply crates/bevy_reflect/src/impls/std.rs:121]: Thank you! That line was the one I understood the least.

Does that mean we should determine a set of applicable traits that Vec<T> reflects and include a `registration.insert` for each of them?

**Comment by CleanCut** [comment]
> Why did you add the `Clone` bound?

Before I committed your suggested changes, I would get the following error without `Clone`:

```
error[E0277]: the trait bound `T: Clone` is not satisfied
   --> crates/bevy_reflect/src/impls/std.rs:118:22
    |
118 |         registration.insert::<Vec<T>>(FromType::<Vec<T>>::from_type());
    |                      ^^^^^^ the trait `Clone` is not implemented for `T`
    |
    = note: required because of the requirements on the impl of `Clone` for `Vec<T>`
    = note: required because of the requirements on the impl of `TypeData` for `Vec<T>`
help: consider further restricting this bound
    |
115 | impl<T: Reflect + Clone> GetTypeRegistration for Vec<T> {
    |                 ^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `bevy_reflect`
```

**Comment by CleanCut** [comment]
@Davier I'm out of my depth here.  I've given you write access to this fork--if you would like to try things out or push your own commits, please feel free to! I'm happy to learn. 😄 

With your changes committed I get:

```
error[E0277]: the trait bound `T: Deserialize<'_>` is not satisfied
   --> crates/bevy_reflect/src/impls/std.rs:118:51
    |
118 |         registration.insert::<ReflectDeserialize>(FromType::<Vec<T>>::from_type());
    |                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `T`
    | 
   ::: crates/bevy_reflect/src/type_registry.rs:224:5
    |
224 |     fn from_type() -> Self;
    |     ----------------------- required by `type_registry::FromType::from_type`
    |
    = note: required because of the requirements on the impl of `for<'a> Deserialize<'a>` for `Vec<T>`
    = note: required because of the requirements on the impl of `type_registry::FromType<Vec<T>>` for `type_registry::ReflectDeserialize`
help: consider further restricting this bound
    |
115 | impl<T: Reflect + Clone + Deserialize<'_>> GetTypeRegistration for Vec<T> {
    |                         ^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

Following the compilers prompts and inferring that `Clone` is unnecessary from your earlier comment, I replaced `Clone` with `Deserialize`, hit another error about lifetimes, and then add a lifetime to Deserialize, (`impl<'a, T: Reflect + Deserialize<'a>> GetTypeRegistration for Vec<T>`) and now I'm stuck on:

```
error: implementation of `Deserialize` is not general enough
   --> crates/bevy_reflect/src/impls/std.rs:118:51
    |
118 |           registration.insert::<ReflectDeserialize>(FromType::<Vec<T>>::from_type());
    |                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ implementation of `Deserialize` is not general enough
    | 
   ::: /Users/nathan/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.117/src/de/mod.rs:531:1
    |
531 | / pub trait Deserialize<'de>: Sized {
532 | |     /// Deserialize this value from the given Serde deserializer.
533 | |     ///
534 | |     /// See the [Implementing `Deserialize`][impl-deserialize] section of the
...   |
569 | |     }
570 | | }
    | |_- trait `Deserialize` defined here
    |
    = note: `Vec<T>` must implement `Deserialize<'0>`, for any lifetime `'0`...
    = note: ...but `Vec<T>` actually implements `Deserialize<'1>`, for some specific lifetime `'1`

error: aborting due to previous error
```

I'm stumped.

**Comment by Davier** [comment]
Looks like it needs a [high rank trait bound](https://doc.rust-lang.org/nomicon/hrtb.html), like there: <https://github.com/bevyengine/bevy/blob/main/crates/bevy_reflect/src/impls/std.rs#L32>. I'm not familiar with it, but I'll investigate this evening.

**Comment by CleanCut** [comment]
Thanks! That did the trick!  I was able to use higher-rank trait bounds with both `Vec<T>` and `HashMap<K, V>` and get them to compile.  I don't know if they actually _work_ at runtime, but given that they compile I've got high hopes. 😄 

`Cow<'static, str>` didn't need a HRTB, since there's no lifetime involved. 

I have never heard of higher-rank trait bounds before today, but after reading that doc and staring at it for awhile it kinda-sorta makes sense.  We're dealing with generics, so we're saying "for _every_ lifetime in this situation" rather than for a specific lifetime.

I should sit down and read through the entire nomicon end-to-end. Time to stop putting it off.

**Pull request merged**